### PR TITLE
Training stabilized

### PR DESCRIPTION
**Pull request by MatteoKartoon**
- learning rate and gradient norm logs in wandb
- learning rate warm up set up
- switch to bfloat16

### CODE REVIEW

File: birefnet/utils.py [hunk 1]
```diff
@@ -30,11 +30,36 @@ def init_wandb(config,args):
...
+        "lr_warm_up_type": config.lr_warm_up_type,
         })
         
         wandb.define_metric("Gradient norm")
         wandb.define_metric("Training Loss")
         wandb.define_metric("Validation Loss")
+        wandb.define_metric("Learning Rate")
+        wandb.define_metric("Gradient norm")
+
+def lr_warm_up(type, epochs, start_factor, end_factor, optimizer):
```
Comment (joelkartoon) [inline birefnet/utils.py:42]: nit: `get_lr_warm_up_scheduler`

**Review by joelkartoon — APPROVED** [review]

File: birefnet/config.py [hunk 2]
```diff
@@ -39,9 +39,10 @@ def __init__(self) -> None:
         self.batch_size = 2
         self.start_epoch=245
         self.log_each_steps = 15
+        self.lr_warm_up_type = 'none'
```
Comment (joelkartoon) [inline birefnet/config.py:42]: use `None` not a string

File: birefnet/loss.py [hunk 3]
```diff
@@ -155,16 +155,18 @@ def __init__(self):
...
             for criterion_name, criterion in self.criterions_last.items():
-                if criterion_name == 'bce' and self.config.bce_with_logits:
-                    _loss = criterion(pred_lvl, gt) * self.lambdas_pix_last[criterion_name]
-                else:
-                    _loss = criterion(pred_lvl.sigmoid(), gt) * self.lambdas_pix_last[criterion_name]
+                if not (criterion_name == 'bce' and self.config.bce_with_logits):
+                    pred_lvl=pred_lvl.sigmoid()
+                loss_crit=criterion(pred_lvl, gt)
+                _loss = loss_crit * self.lambdas_pix_last[criterion_name]
+                loss_components[criterion_name]+=loss_crit.detach()
```
Comment (joelkartoon) [inline birefnet/loss.py:167]: should probably be `_loss` not `loss_crit`

File: birefnet/loss.py [hunk 4]
```diff
@@ -155,16 +155,18 @@ def __init__(self):
 
     def forward(self, scaled_preds, gt):
         loss = 0.
+        loss_components={k:0. for k in self.criterions_last.keys() if self.lambdas_pix_last[k]}
```
Comment (joelkartoon) [inline birefnet/loss.py:158]: use the dictionary which uses boolean values

File: scripts/train.py [hunk 5]
```diff
@@ -264,6 +294,7 @@ def _batch(self, batch, batch_idx, epoch, validation=False):
         self.optimizer.zero_grad()
         scaled_preds, class_preds_lst = self.model(inputs)
         loss_dict=self.loss_dict_validation if validation else self.loss_dict_train
+        loss_components=self.loss_components_validation if validation else self.loss_components_train
```
Comment (joelkartoon) [inline scripts/train.py:297]: unused

File: scripts/train.py [hunk 6]
```diff
@@ -281,7 +312,11 @@ def _batch(self, batch, batch_idx, epoch, validation=False):
             loss_dict['loss_cls'] = loss_cls.item()
         
         # Loss
-        loss_pix = self.pix_loss(scaled_preds, torch.clamp(gts, 0, 1)) * 1.0
+        loss_pix, loss_components = self.pix_loss(scaled_preds, torch.clamp(gts, 0, 1))
+        if not validation:
```
Comment (joelkartoon) [inline scripts/train.py:316]: avoid unnecessary negations

File: scripts/train.py [hunk 7]
```diff
@@ -325,6 +361,8 @@ def train_epoch(self, epoch):
         self.model.train()
         self.loss_dict_train = {}
         self.loss_dict_validation = {}
+        self.loss_components_train = {}
```
Comment (joelkartoon) [inline scripts/train.py:364]: why do we need this?

**Review by joelkartoon — APPROVED** [review]
